xmipptomo/protocols/protocol_extract_subtomos.py

```diff
@@ -0,0 +1,222 @@
+# *
+# **************************************************************************
+
+import os
+import glob
+from pwem.emlib import lib
+from pwem.objects import Particle, Volume, Transform, String, SetOfVolumes, SetOfParticles
+from pwem.protocols import EMProtocol
+from pwem import ALIGN_PROJ
+import pwem.emlib.metadata as md
+
+from pyworkflow import BETA
+from pyworkflow import utils as pwutils
+from pyworkflow.protocol.params import PointerParam, EnumParam, IntParam, BooleanParam
+
+from tomo.objects import SetOfTomograms, SetOfSubTomograms, SubTomogram, SetOfCoordinates3D, TomoAcquisition, MATRIX_CONVERSION
+import tomo.constants as const
+
+from tomo.protocols import ProtTomoBase
+from xmipp3.convert import alignmentToRow
+
+
+COORD_BASE_FN = 'coords'
+# Tomogram type constants for particle extraction
+SAME_AS_PICKING = 0
+OTHER = 1
+
+
+class XmippProtExtractSubtomos(EMProtocol, ProtTomoBase):
+    """
+    Extract a set of subtomograms from a set of tomograms given a set of coordinates.
+    """
+    _label = 'extract subtomos'
+    _devStatus = BETA
+    #_possibleOutputs = SetOfSubTomograms
+    #OUTPUT_PREFIX = _possibleOutputs.subtomograms.name
+    lines = []
+    tomoFiles = []
+
+    # --------------------------- DEFINE param functions ------------------------
+    def _defineParams(self, form):
+        form.addSection(label='Parameters')
+
+        form.addParam('tomograms', PointerParam, pointerClass=SetOfTomograms,
+                      label='Tomograms', help='The subtomograms will be extracted from this set.')
+
+        form.addParam('coords', PointerParam, pointerClass=SetOfCoordinates3D,
+                  label='Coordinates', help='The coordinates that definethe position of the set of subtomograms to be extracted.'
+                                            'The coordinate denotes the center of the subtomogram')
+
+        form.addParam('boxSize', IntParam,
+                      label='Box size',
+                      help='The subtomograms are extracted as a cube. The box size defines the edge of the cube'
+                           'The wizard selects same box size as picking')
+
+        form.addParam('invertContrast', BooleanParam,
+                      label='Invert Contrast',  default=True,
+                      help='Normally, tomograms has the contrast inverted with respect to standard for subtomograms. '
+                           'It means, in the tomograms the structure is black and the noise is white. Generally, the'
+                           'subtomograms are white with a black background. This means that the subtomograms has the '
+                           'contrast inverted with respect to the tomograms. Put this flag as True if the contrast'
+                           'need to be inverted.')
+
+    # --------------------------- INSERT steps functions ------------------------
+    def _insertAllSteps(self):
+
+        for tom in self.tomograms.get():
```

Comment:
I think this is better to get only the tomograms present in the coordinates in case there is a subset of 3d coordinates.

Use --> https://github.com/scipion-em/scipion-em-tomo/blob/devel/tomo/objects.py#L1501